external-crates/move/crates/move-analyzer/src/completion.rs

```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
+        .values()
+        .flatten()
+        .find(|mdef| mdef.ident == mod_ident.value)
+    else {
+        return completions;
+    };
+
+    let Some(edef) = mod_defs.enums.get(&datatype_name) else {
+        return completions;
+    };
+
+    let Some(DefInfo::Enum(.., variants, _)) = symbols.def_info.get(&edef.name_loc) else {
+        return completions;
+    };
+
+    for vinfo in variants {
+        let Some(DefInfo::Variant(_, _, vname, is_positional, field_names, ..)) =
+            symbols.def_info.get(&vinfo.name.loc)
+        else {
+            continue;
+        };
+
+        let label = if field_names.is_empty() {
+            vname.to_string()
+        } else if *is_positional {
+            format!("{vname}()")
+        } else {
+            format!("{vname}{{}}")
+        };
+        let field_snippet = field_names
+            .iter()
+            .enumerate()
+            .map(|(snippet_idx, fname)| {
+                if *is_positional {
+                    format!("${{{}}}", snippet_idx + 1)
+                } else {
+                    format!("${{{}:{}}}", snippet_idx + 1, fname)
+                }
+            })
+            .collect::<Vec<_>>()
+            .join(", ");
+        let insert_text = if *is_positional {
+            format!("{vname}({field_snippet})")
+        } else {
+            format!("{vname}{{{field_snippet}}}")
+        };
+
+        completions.push(CompletionItem {
+            label,
+            kind: Some(CompletionItemKind::ENUM_MEMBER),
+            insert_text: Some(insert_text),
+            insert_text_format: Some(InsertTextFormat::SNIPPET),
+            ..Default::default()
+        });
+    }
+
+    completions
+}
+
+/// Computes completions for access path entries (components beyond the leading name)
+fn entries_completions(
+    symbols: &Symbols,
+    cursor: &CursorContext,
+    info: &AliasAutocompleteInfo,
+    prev_info: ChainComponentInfo,
+    chain_target: ChainCompletionTarget,
+    path_entries: &[Name],
+    path_index: usize,
+    colon_colon_triggered: bool,
+    completions: &mut Vec<CompletionItem>,
+) {
+    let ChainComponentInfo {
+        loc: prev_loc,
+        kind: prev_kind,
+    } = prev_info;
+
+    let mut at_colon_colon = false;
+    if path_index == path_entries.len() {
+        // the only reason we would not return here is if we were at `::` which is past the location
+        // of the last path component
+        if colon_colon_triggered && cursor.loc.start() > prev_loc.end() {
+            at_colon_colon = true;
+        } else {
+            return;
+        }
+    }
+
+    if !at_colon_colon {
+        // we are not at the last `::` but we may be at an intermediate one
+        if colon_colon_triggered
+            && path_index < path_entries.len()
+            && cursor.loc.start() > prev_loc.end()
+            && cursor.loc.end() <= path_entries[path_index].loc.start()
+        {
+            at_colon_colon = true;
+        }
+    }
```

Comment:
Because we first need to check if we are at the last `::` (and return if we are not) and only then check if we at an intermediate one, to me it feels like pulling this into its own function would result in code that's more confusing